File: src/tokens.zig
pub const char = u32;
pub const Tag = union(enum) {
    // Punctuation
    LBrace: void,
    RBrace: void,
    LBracket: void,
    RBracket: void,
    LParen: void,
    RParen: void,
    Store: void, // ->
    Colon: void,
    Semicolon: void,
    Comma: void,
    Dot: void,

    ///// OPERATORS

    // Assertions
    // These are parsed at the same level as assignments
    Assert: void, // 'assert'

    // Assignment Operators
    Assign: void,
    AddAssign: void,
    SubAssign: void,
    MulAssign: void,
    DivAssign: void,
    ModAssign: void,
    ShlAssign: void,
    ShrAssign: void,

    // Standard expression ops
    Add: void,
    Sub: void,
    Mul: void,
    Div: void,
    Mod: void,

    // Logical
    Not: void,
    And: void,
    Or: void,
    Xor: void,

    // Relational
    Eq: void,
    NotEq: void,
    Less: void,
    Greater: void,
    LessEq: void,
    GreaterEq: void,

    // Range
    In: void,
    OpenRange: void,
    ClosedRange: void,

    // Bit twiddlers
    BitNot: void,
    BitAnd: void,
    BitOr: void,
    BitXor: void,

    // Code-mods
    Array: void, // These 2 since [] is ambiguous with a compound literal.
    Slice: void,
    Comptime: void,
    Pure: void,
    Inline: void,
    Optional: void, // ?
    // Pointer should also be here, but it's the same as Mul

    // Binds
    Let: void,
    Var: void,
    CVar: void,
    Field: void,
    Enum: void,
    Property: void,
    Alias: void, // New one: Creates a simple alias. The AST equivalent of a pointer

    // Typedefs
    StructDef: void,
    EnumDef: void,

    // Control Flow
    If: void,
    ElIf: void,
    Else: void,
    For: void,
    While: void,
    Loop: void,
    Finally: void,
    Switch: void, // Not specced yet, but planned.
    Fn: void,
    Break: void,
    Return: void,

    Label: []const u8,
    Symbol: []const u8,
    Sink: void, // '_'

    // Literals
    StringLit: []const u8, // This string is allocated apart from the file (TODO)
    CharLit: char,
    BoolLit: bool, // Reserves both 'true' and 'false'
    IntLit: usize, // If it can't fit in a usize, it shouldn't be a literal.
    FloatLit: f64, // See above
};

// So we don't have to declare all those tags twice
pub const TagType = @TagType(Tag);

pub const FilePos = struct {
    file_id: usize,
    line: usize,
    col: usize,
};
pub const Token = struct {
    pos: FilePos,
    tag: Tag,
};

File: src/tokens.nim
import strutils
import tags



# Expression hierarchy is Binary->Unary->Access->(Repeat in parens)


type BinLevel*{.pure.} = enum
  Assignment, Equality, Relational, Ors, Ands, Membership, Range, Arithmetic, Product, Bitwise

template below*(level: BinLevel): BinLevel =
  BinLevel(ord(level) + 1)

# All expression-level binary operators.
# FieldAccess and such are not included here as they are parsed below even unary
const BinOps*: array[BinLevel, set[Tag]] = [
  {Tag.Assign, Tag.AddAssign, Tag.SubAssign, Tag.MulAssign, Tag.DivAssign, Tag.BitOrAssign, Tag.BitAndAssign, Tag.ShlAssign, Tag.ShrAssign},
  {Tag.Equal, Tag.NotEqual},
  {Tag.Less, Tag.Greater, Tag.GreaterEq, Tag.LessEq, Tag.Spaceship},
  {Tag.Or, Tag.Xor},
  {Tag.And},
  {Tag.In, Tag.NotIn},
  {Tag.OpenRange, Tag.ClosedRange},
  {Tag.Add, Tag.Sub},
  {Tag.Mul, Tag.Div, Tag.Mod},
  {Tag.BitOr, Tag.BitAnd, Tag.BitXor},
]

  

const UnaryOps*: set[Tag] = {
  Tag.Sub, Tag.BitNot, Tag.Not,
  Tag.Const, Tag.Comptime, # Used for type expressions
  Tag.Array, Tag.Slice, Tag.Optional,
  Tag.Pure, Tag.Inline, Tag.Struct, Tag.Enum, Tag.Overload, Tag.Property,
  # As pointer
  Tag.Mul
}

const CallOps*: set[Tag] = { Tag.LParen, Tag.LBracket }

const BindSpecs*: set[Tag] = {
  Tag.Let, Tag.Var, Tag.CVar, Tag.Field, Tag.Enum
}

# Note that tuples are technically atoms, though they aren't included here
const Atoms*: set[Tag] = {
  Tag.Symbol, Tag.NullLit, Tag.IntLit, Tag.StringLit
}
const Literals*: set[Tag] = {
  Tag.NullLit, Tag.IntLit, Tag.StringLit
}

const Closers*: set[Tag] = {
    Tag.RParen, Tag.RBracket, Tag.RBrace
}

# In foo.bar.baz, these are the bar/baz
const ValidSwizzles*: set[Tag] = {
  Tag.Symbol, Tag.IntLit
}

const ValidSymbolBeginnings*: set[char] = {'_', '@'} + Letters
const ValidSymbolChars*: set[char] = ValidSymbolBeginnings + Digits
# A stroke of insanity: I'll now parse float lits as field accesses into ints

type
  FilePos* = tuple[line:uint, col:uint]
  Token* = object
    pos*: FilePos
    case tag*: Tag
    of Tag.Symbol, Tag.Label,  StringLit, CharLit:
      lexeme*: string
    of Tag.IntLit:
      val*: uint
    else: discard

